### PR TITLE
 #11 Introduce TripViewPage to see the trips, edit start & end time

### PR DESCRIPTION
**Pull request by MxmIv**
Introduce TripViewPage to see the trips, edit start & end time, see stops for the trip + Add Unit Tests for TripItemViewModel

Testing: 
- Pulled all the recent changed from `dev` branch and build 
- Tested locally
- Added unit tests (there is a failing test not related to the change) 
Trip1Details

The whole flow of the app can be seen here: 

- Updating time of the completed trip 
- Updating time of the ongoing trip and marking it complete 

Trip1Details
Trip12Details
Trip12UpdatedDetails
Trip1UpdatedDetails

### CODE REVIEW

**Comment by FMulligan** [comment]
Build seems to have failed due to changed DbContext

File: HaulageApplication/HaulageApp/ViewModels/TripItemViewModel.cs [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        }
+
+        public event PropertyChangedEventHandler? PropertyChanged;
+
+        protected void OnPropertyChanged(string propertyName)
+        {
+            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
+        }
+    }
+}
```
Comment (FMulligan) [inline HaulageApplication/HaulageApp/ViewModels/TripItemViewModel.cs:163]: Is there a reason to not make the class an Observable Object which would already have OnPropertyChanged implemented?

File: HaulageApplication/HaulageApp/ViewModels/TripViewModel.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+                }
+            }
+        }
+
+        public event PropertyChangedEventHandler? PropertyChanged;
+
+        protected void OnPropertyChanged(string propertyName)
+        {
+            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
+        }
```
Comment (FMulligan) [inline HaulageApplication/HaulageApp/ViewModels/TripViewModel.cs:56]: Same question as previous

File: HaulageApplication/HaulageApp/Models/Vehicle.cs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+namespace HaulageApp.Models
+{
+    public class Vehicle
+    {
+        public int VehicleId { get; set; }
+        public string VehicleType { get; set; }
+        public int Capacity { get; set; }
+        public string Status { get; set; }
```
Comment (FMulligan) [inline HaulageApplication/HaulageApp/Models/Vehicle.cs]: Needs to be rebased inline with pushed changes

File: HaulageApplication/HaulageApp/ViewModels/TripItemViewModel.cs (hunk 1 above)
Comment (MxmIv) [reply HaulageApplication/HaulageApp/ViewModels/TripItemViewModel.cs:163]: Implementing INotifyPropertyChanged is needed ensure that the Date and Time are fully entered before triggering a property change.

In previous attempts, I encountered issues with premature validation, where changes were being applied before the input was fully completed. This manual approach allows me to have precise control over when the OnPropertyChanged method is called -> after the input is complete

File: HaulageApplication/HaulageApp/Models/Waypoint.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        [ForeignKey("trip")]
+        [Column("trip_id")] 
+        public int TripId { get; set; }
+
+        [ForeignKey("user")]
+        [Column("user_id")] 
+        public int UserId { get; set; }
+
+        [Column("location")]  
+        public string Location { get; set; }
```
Comment (FMulligan) [inline HaulageApplication/HaulageApp/Models/Waypoint.cs:22]: [Nit] Location, WaypointType & Trip could be set as required

File: HaulageApplication/HaulageApp/ViewModels/TripViewModel.cs (hunk 2 above)
Comment (MxmIv) [reply HaulageApplication/HaulageApp/ViewModels/TripViewModel.cs:56]: same issue as above

File: HaulageApplication/HaulageApp/Models/Vehicle.cs (hunk 3 above)
Comment (MxmIv) [reply HaulageApplication/HaulageApp/Models/Vehicle.cs]: understood

File: HaulageApplication/HaulageApp/Models/Trip.cs [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+        public int Driver { get; set; } 
+
+        [Column("start_time")]
+        public DateTime StartTime { get; set; }
+
+        [Column("end_time")]
+        public DateTime? EndTime { get; set; } 
+
+        [Column("status")]
+        public string Status { get; set; }
```
Comment (FMulligan) [inline HaulageApplication/HaulageApp/Models/Trip.cs:24]: [Nit] Status & Waypoints could be set as required

**Review by FMulligan — APPROVED** [review]
With the Rebased Changes - looks good!

**Pull request merged**